mantidproject/mantid
# Extend trans command to allow beam stop out method

Fixes #10192

**Context**

The main part and functionality of the code was implemented by @PeterParker . 
An performance analysis of the code showed a bottleneck when extracting spectra with the help of detector IDs. Moving this functionality to the C++ side of things improved this bottleneck considerably (~300x).
### For Testing

The functionality was tested by Richard. Hence we will only perform functional checks here

The required data and user files can be found here: \olympic\Babylon5\Scratch\Anton\Testing\10192_Extend_Trans_Command_For_Beam_Stop
You need to make sure that the path to the data and the output path are set correctly. 
-  The instrument is SANS2DTUBES
  Setup for "Test old TRANSPEC setting":
  Sample:
-  The scattering sample is SANS2D00023785-add.nxs
-  The transmission sample is SANS2D00023775.nxs
-  The direct sample is  SANS2D00023774.nxs

Can:
-  The scattering can is SANS2D00023784-add.nxs
-  The transmission can is SANS2D00023774.nxs
-  The direct can is SANS2D00023774.nxs

Setup for for all other tests:
Sample:
-  The scattering sample is SANS2D00023785-add.nxs
-  The transmission sample is SANS2D00023763-add.nxs
-  The direct sample is  SANS2D00023764-add.nxs

Can:
-  The scattering can is SANS2D00023784-add.nxs
-  The transmission can is SANS2D00023764-add.nxs
-  The direct can is SANS2D00023764-add.nxs

**Test old TRANSPEC setting**

The user file does not use the beam stop for its calculation, ie TRANS/TRANSPEC=3
1. Open the ISIS SANS GUI and make sure that the instrument is set.
2. Load the user file: user_file_M3_old.txt
3. Populate the data entries with the inputs specified above. Press load
4. Press 1D Reduce
   - Confirm that no errors appear

**Test ROI setting**

We want to make sure that we can set a ROI by specifying a mask file. Additionally we want to exclude some areas from the ROI. The user file contains the commands TRANS/ROI=TRANS_ROI.xml and TRANS/MASK=TRANS_MASK.xml.
1. Open the ISIS SANS GUI and make sure that the instrument is set.
2. Load the user file: user_file_with_bs_ROI.txt
3. Populate the data entries with the inputs specified above. Press load
4. Press 1D Reduce
   - Confirm that no errors appear

**Test Radius setting**

We want to make sure that we can set a Radius on the beam stop. The user file contains
TRANS/RADIUS=30
1. Open the ISIS SANS GUI and make sure that the instrument is set.
2. Load the user file: user_file_with_bs_RADIUS.txt
3. Populate the data entries with the inputs specified above. Press load
4. Press 1D Reduce
   - Confirm that no errors appear

**Test combination radius and roi file**

We want to make sure that we can set a mix of Radius and ROI.
1. Open the ISIS SANS GUI and make sure that the instrument is set.
2. Load the user file: user_file_with_bs_ROI_and_RADIUS.txt
3. Populate the data entries with the inputs specified above. Press load
4. Press 1D Reduce
   - Confirm that no errors appear

**Test combination radius and roi file and monitor**

We want to make sure that we can set a mix of Radius and ROI and also setting M3. Note that Setting a monitor for transmission takes precedence.
1. Open the ISIS SANS GUI and make sure that the instrument is set.
2. Load the user file: user_file_with_bs_ROI_and_RADIUS_and_Monitor.txt
3. Populate the data entries with the inputs specified above. Press load
4. Press 1D Reduce
   - Confirm that no errors appear

When this passes testing, please inform @picatess in order to update the SANS Wiki.
